lib/srv/db/redis/engine.go

```diff
@@ -195,7 +196,18 @@ func (e *Engine) getNewClientFn(ctx context.Context, sessionCtx *common.Session)
 	}
 
 	return func(username, password string) (redis.UniversalClient, error) {
-		redisClient, err := newClient(ctx, connectionOptions, tlsConfig, username, password)
+		var onConnect onClientConnectFunc
+		switch {
+		// If password is provided by client.
+		case password != "":
+			onConnect = authWithPasswordOnConnect(username, password)
+
+		// If database user is one of managed users.
+		case apiutils.SliceContainsStr(sessionCtx.Database.GetManagedUsers(), sessionCtx.DatabaseUser):
```

Comment:
Yes. That is true for non-managed users. But I think it's the same for self-hosted Redis.

`default` user is used right after the connection until `auth <xxx> <xxx>` is manually input from redis-cli.

Reply:
During the initial client connection, if you only pass in a username without any password, nothing would happen. I think it will still default to the `default` user's permission.

On way to ensure the correct user, we can wait for `auth` command and block all other commands. The engine has to keep a state like `waitingForAuth`. @jakule @smallinsky what do you think?

Reply:
as discussed, this is the default behaviour for Redis's sessions that the default user is used until a proper `auth` command is sent. It's best for DB admins to configure proper access for the default user.

Teleport currently makes sure that.
- for non-managed users, the user in the `auth` command must match `--db-user` from tsh command
- for managed users, `auth` commands are ignored. 

I think it is possible for Teleport to block any commands other than `auth` if `--db-user` is specified but not sure if that's worth it. If we would like to enforce that way, I can create an issue and track it separately as it's tied to ElastiCache per say.